Compatibility-2.0.md

```diff
@@ -26,3 +26,304 @@ Embedded Mode also might use a decorated Cache using a Transcoder. For example t
+
+Example:
+```java
+// Bypass all encoding and put in the stored format directly:
+cache.getAdvancedCache().withEncoding(IdentityEncoder.class).put(...)
+
+// Write a json
+// Will cause a transcoding from json to xml first and then it will be stored directly
+cache.getAdvancedCache().withMimeType("application/json").withEncoding(IdentityEncoder.class).put("{json}")
+```
+
+##### Reading via REST
+
+The rest client will include an Accept header asking for the format it wants the data to be returned:
+
+```
+curl -X GET -H 'Accept: application/json' http://localhost:8080/rest/cache/1
+```
+
+Internally, in the server, this will involve the following operations:
+
+```
+Stored format ----(Encoder.fromStorage)---->  XML (string or byte[])  ----(JsonXMLTranscoder.transcode)----> Json (String)  ---> Client
+```
+
+##### Reading from Hot Rod
+
+Clients will receive the ```byte[]``` back, as Hot Rod does not allow to request data in a specific format (yet) so we initially assume the data will be received without 
+any transcoding
+
+##### Reading from embedded
+
+Same as for writing, embedded caches are free to specify any encoding on the fly
+
+#### 2. Querying and Indexing using json
+
+Infinispan's query engine supports two data formats: protobuf and java entities annotated with hibernate search.
+By using transcoding capabilities, we can enable querying and indexing by dealing with only json data in the caller:
+
+Example configuration:
+
+```xml
+<local-cache>
+  <data-type>
+     <!-- No encoder declared, so it will store whatever is passed in -->
+     <key mime-type="application/x-protostream"/>
+     <value mime-type="application/x-protostream"/>
+  </data-type>
+</local-cache>
+```
+
+When a request arrives to the REST endpoint in the form:
+
+```
+curl -X POST -H 'Content-type: application/json' -d @file.json http://localhost:8080/rest/cache/1
+```
+
+Infinispan will lookup for a transcoder that can convert between ```application/x-protostream``` and ```application/json```
+
+and will pass the converted result (a protobuf) to the cache, enabling thus, the indexing to happen.
+
+
+WARNING: It'll be slightly more complicated than that, as first we need to craft a .proto file for the server and register 
+it.
+
+At the same time, Hot Rod clients would deal with protostream directly. They will probably have message marshallers to convert to and 
+from protobuf.  Once Hot Rod has the capability of asking for a particular MimeType, they would be able to get the json directly.
+
+
+
+#### 3. Compressing server
+
+It'd be also possible to transparently compress and de-compress data by using the right configuration:
+
+```xml
+<local-cache>
+  <data-type>
+     <!-- keys are Strings, values are text that will be stored compressed -->
+     <key mime-type="application/x-java-object"/>
+     <value mime-type="text/plain" encoder="org.infinispan.commons.dataconversion.GzipEncoder"/>
```

Comment:
I am slightly confused about the mime-type with encoder: is the mime-type expected to be input format for the encoder, or its output format? Because GzipEncoder would transform text/plain -> application/octet-stream or [application/gzip](https://www.iana.org/assignments/media-types/application/gzip)